ui/src/treegrid.tsx

```diff
@@ -0,0 +1,247 @@
+import ErrorIcon from "@mui/icons-material/Error";
+import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
+import KeyboardArrowRightIcon from "@mui/icons-material/KeyboardArrowRight";
+import CircularProgress from "@mui/material/CircularProgress";
+import IconButton from "@mui/material/IconButton";
+import Typography from "@mui/material/Typography";
+import { ReactNode, useCallback } from "react";
+import { useImmer } from "use-immer";
+
+export type TreeGridId = string | number;
+
+export type TreeGridRowData = {
+  [key: string]: string | number | JSX.Element;
+};
+
+export type TreeGridItem = {
+  children?: TreeGridId[];
+  data: TreeGridRowData;
+};
+
+export type TreeGridData = {
+  [key: TreeGridId]: TreeGridItem;
+};
+
+export type TreeGridColumnData = {
+  key: string;
+  width: string | number;
+  title?: string | JSX.Element;
+};
+
+export type TreeGridProps = {
+  columns: TreeGridColumnData[];
+  data: TreeGridData;
+  prefixElements?: (id: TreeGridId, data: TreeGridRowData) => ReactNode;
+  loadChildren?: (id: TreeGridId) => Promise<string>;
+};
+
+export function TreeGrid(props: TreeGridProps) {
+  const [state, updateState] = useImmer<TreeGridState>(
```

Comment:
I'm not familiar with immer, though seems reasonable at a glance - how did you land on this? Any alternatives considered?

Reply:
I was unsatisfied with the lodash deepCopy method I had been using so far. I didn't find any other simple options beyond using basic javascript (e.g. spreading) which I've found unwieldy with nontrivial state. I'm also considering whether it makes sense to buy into a larger ecosystem like Redux or Recoil. Conveniently, I'm leaning towards Redux which uses immer so this works well both independently and also if I go that route. The other options for this used by Recoil and other systems generally aren't separable from their systems.